src/lib/twizzler-driver/src/dma/object.rs

```diff
@@ -1,46 +1,95 @@
-use twizzler_object::Object;
+use std::sync::Mutex;
 
-use super::{Access, DeviceSync, DmaArrayRegion, DmaOptions, DmaRegion};
+use twizzler_abi::{
+    kso::{KactionCmd, KactionFlags, KactionGenericCmd},
+    object::{MAX_SIZE, NULLPAGE_SIZE},
+    syscall::sys_kaction,
+};
+use twizzler_object::{ObjID, Object};
 
+use super::{Access, DeviceSync, DmaOptions, DmaRegion, DmaSliceRegion};
+
+/// A handle for an object that can be used to perform DMA, and is most useful directly as a way to
+/// perform DMA operations on a specific object. For an allocator-like DMA interface, see [crate::pool::DmaPool].
 pub struct DmaObject {
     obj: Object<()>,
+    pub(crate) releasable_pins: Mutex<Vec<u32>>,
 }
 
 impl DmaObject {
+    /// Create a [DmaSliceRegion] from the base of this object, where the region represents memory
+    /// of type `[T; len]`.
     pub fn slice_region<'a, T: DeviceSync>(
         &'a self,
         len: usize,
         access: Access,
         options: DmaOptions,
-    ) -> DmaArrayRegion<'a, T> {
-        todo!()
+    ) -> DmaSliceRegion<'a, T> {
+        let nr_bytes = core::mem::size_of::<T>()
+            .checked_mul(len)
+            .expect("Value of len too large");
+        assert!(nr_bytes < MAX_SIZE - NULLPAGE_SIZE * 2);
```

Comment:
I know it's nit picky but I usually parenthesize the operations to show the order of them rather than depending on my memory.

Reply:
Are you referring to the < vs - precedence? Rust wouldn't allow `(nr_bytes < MAX_SIZE) - NULLPAGE_SIZE * 2`, as that wouldn't type-check.

Reply:
Yeah, I'm probably still reading Rust as C too much.